File: 20 risk - external tools, bracket trailing/cpp/bracket_trailing.cpp
#include "tse_helpers.hpp"

#include <cstdio>
#include <string>
#include <vector>

int main()
{
	tse::setLogLevel(tse::LogLevel::none);

	tse::Account account {"RiskManagementExternalTools BracketTrailing", tse::StorageRegime::mem};
	std::vector<tse::OhlcvTick> const ticks {helpers::loadAapl()};

	tse::Market const market {account.createMarket("OHLCV", tse::MdType::ohlcv)};
	account.createSimulator("Sim", helpers::simulatorConfig(), 200, -1);
	account.addContract("AAPL", 1, tse::Instrument::equity, tse::Underlying::undefined, tse::Venue::undefined, 100000);

	account.addInputOhlcv("AAPL SMA fast", 50, tse::Duration::days, helpers::makeSma(50), market, {"AAPL"});
	account.addInputOhlcv("AAPL SMA slow", 200, tse::Duration::days, helpers::makeSma(200), market, {"AAPL"});
	account.addPatternCrossover("PatternToLong", tse::Duration::days, {"AAPL SMA fast", "AAPL SMA slow"}, tse::Cmp::ge);

	// Buy 100 shares when the fifty-day average crosses above the two-hundred-day one.
	tse::LegDescriptor const entry
	{
		"AAPL",
		tse::QuantityMode::fixed,
		100.0,
		tse::PriceType::market,
		150.0,
		0.0,
		0.0,
		tse::TxnType::enter,
		tse::Side::long_,
		tse::Side::neutral,
		tse::Tif::day,
		tse::Priority::replaceable
	};
	// Both legs rest at the venue again, but the stop now trails: it follows the highest price the
	// trade has reached, five percent behind, so gains already made are not handed back. The ten
	// percent target stays where the fill put it.
	tse::VenueRiskSpec const risk
	{
		tse::VenueRiskLeg {true, 1, 0.05},
		tse::VenueRiskLeg {true, 0, 0.10},
		tse::Tif::gtc
	};
	account.addRuleBracket("Bracket", entry, risk, "PatternToLong");
	account.addRobot("Bracket robot", {"Bracket"});
	account.start("Bracket robot");

	for (tse::OhlcvTick const& tick : ticks) {
		market.pushOhlcv("AAPL", tick);
	}

	tse::Summary const summary {account.getSummary()};
	std::printf
	(
		"bracket trailing: trades=%lld netProfit=%.4f grossProfit=%.4f grossLoss=%.4f profitFactor=%.6f maxDrawdown=%.4f\n",
		static_cast<long long>(summary.totalNumberOfTrades),
		summary.totalNetProfit,
		summary.grossProfit,
		summary.grossLoss,
		summary.profitFactor,
		summary.maxDrawdown
	);
	return 0;
}

File: 20 risk - external tools, bracket trailing/python/bracket_trailing.py
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "helpers", "python"))
import tse_helpers as H

tse = H.tse


def main():
    account = H.account("RiskManagementExternalTools BracketTrailing", tse.StorageRegime.Mem, tse.LogLevel.Off)
    ticks = H.load_aapl()

    market = account.create_market("OHLCV", tse.MdType.Ohlcv)
    account.create_simulator("Sim", H.simulator_options(), 200, -1)
    account.add_contract("AAPL", 1, tse.Instrument.Equity, tse.Underlying.Undefined, tse.Venue.Undefined, 100000)

    account.add_input_ohlcv("AAPL SMA fast", 50, tse.Duration.Days, H.make_sma(50), market, ["AAPL"])
    account.add_input_ohlcv("AAPL SMA slow", 200, tse.Duration.Days, H.make_sma(200), market, ["AAPL"])
    account.add_pattern_crossover("PatternToLong", tse.Duration.Days, ["AAPL SMA fast", "AAPL SMA slow"], tse.Cmp.Ge)

    # Buy 100 shares when the fifty-day average crosses above the two-hundred-day one.
    entry = tse.make_leg_descriptor(
        "AAPL",
        tse.Quantity.Fixed,
        100.0,
        tse.Price.Market,
        150.0,
        0.0,
        0.0,
        tse.TxnType.Enter,
        tse.Side.Long,
        tse.Side.Neutral,
        tse.Tif.Day,
        tse.Priority.Replaceable,
    )
    # Both legs rest at the venue again, but the stop now trails: it follows the highest price the
    # trade has reached, five percent behind, so gains already made are not handed back. The ten
    # percent target stays where the fill put it.
    risk = tse.make_venue_risk_spec(
        tse.make_venue_risk_leg(True, 1, 0.05),
        tse.make_venue_risk_leg(True, 0, 0.10),
        tse.Tif.Gtc,
    )
    account.add_rule_bracket("Bracket", entry, risk, "PatternToLong")
    account.add_robot("Bracket robot", ["Bracket"])
    account.start("Bracket robot")

    for tick in ticks:
        market.push_ohlcv_by_name("AAPL", tick)

    summary = account.get_summary()
    account.close()

    print("bracket trailing: trades={} netProfit={:.4f} grossProfit={:.4f} grossLoss={:.4f} profitFactor={:.6f} maxDrawdown={:.4f}".format(
        summary.totalNumberOfTrades, summary.totalNetProfit, summary.grossProfit,
        summary.grossLoss, summary.profitFactor, summary.maxDrawdown))
    return 0


if __name__ == "__main__":
    sys.exit(main())
